graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,327 @@
+
+      if (directive.isPresent()) {
+        fieldPolicy.replace(directive.get());
+        inheritMaxAge = directive.get().getInheritMaxAge();
+      }
+    }
+
+    // Cache directive on the field itself
+
+    Optional<CacheControlDirective> fieldDirective = CacheControlDirective.fromDirectiveContainer(parameters.getField());
+    if (fieldDirective.isPresent()) {
+      CacheControlDirective directive = fieldDirective.get();
+
+      // If inheritMaxAge is true, take note of that to avoid setting the
+      // default max age in the next step. This does allow setting the cache
+      // scope though.
+      //
+      // Note: setting both maxAge and inheritMaxAge doesn't make sense.
+      if (directive.getInheritMaxAge() && !fieldPolicy.hasMaxAge()) {
+        inheritMaxAge = true;
+        fieldPolicy.replace(directive.getScope());
+      } else {
+        fieldPolicy.replace(directive);
+      }
+    }
+
+    // If this field returns a composite type or is a root field and
+    // we haven't seen an explicit maxAge argument, set the maxAge to 0
+    // (uncached) or the default if specified in the constructor.
+    // (Non-object fields by default are assumed to inherit their
+    // cacheability from their parents. But on the other hand, while
+    // root non-object fields can get explicit directives from their
+    // definition on the Query/Mutation object, if that doesn't exist
+    // then there's no parent field that would assign the default
+    // maxAge, so we do it here.)
+    //
+    // You can disable this on a non-root field by writing
+    // `@cacheControl(inheritMaxAge: true)` on it. If you do this,
+    // then its children will be treated like root paths, since there
+    // is no parent maxAge to inherit.
+
+    if (!fieldPolicy.hasMaxAge()
+        && ((unwrappedFieldType instanceof GraphQLCompositeType && !inheritMaxAge)
+        || parameters.getExecutionStepInfo().getPath().isRootPath())) {
+      fieldPolicy.restrict(defaultMaxAge);
+    }
+
+    return new SimpleInstrumentationContext<ExecutionResult>() {
+      @Override
+      public void onCompleted(ExecutionResult result, Throwable t) {
+        state.overallPolicy.restrict(fieldPolicy);
+        super.onCompleted(result, t);
+      }
+    };
+  }
+
+  enum CacheControlScope {
+    PUBLIC,
+    PRIVATE
+  }
+
+  private static class CacheControlState implements InstrumentationState {
+    public final CacheControlPolicy overallPolicy = new CacheControlPolicy();
+  }
+
+  private static class CacheControlPolicy {
+    @Nullable
+    private Integer maxAge;
+    @Nullable
+    private CacheControlScope scope = CacheControlScope.PUBLIC;
+
+    void restrict(CacheControlPolicy policy) {
+      if (policy.maxAge != null && (maxAge == null || policy.maxAge < maxAge)) {
+        this.maxAge = policy.maxAge;
+      }
+
+      if (policy.scope != null && (scope == null || !scope.equals(CacheControlScope.PRIVATE))) {
+        this.scope = policy.scope;
+      }
```

Comment:
Nit: because you have individual, `restrict(maxAge)` and `restrict(scope)` below, I would certainly make them accept nullable values and rewrite this as:

```java
    void restrict(CacheControlPolicy policy) {
      restrict(policy.maxAge)
      restrict(policy.scope)
}
```